File: src/test/java/com/github/natche/gravatarjavaclient/profile/GravatarProfileUrlTest.java
package com.github.natche.gravatarjavaclient.profile;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for {@link GravatarProfileUrl}s.
 */
class GravatarProfileUrlTest {
    /**
     * Creates a new instance of this class for testing purposes.
     */
    GravatarProfileUrlTest() {}

    /**
     * Tests for creation of Gravatar profile urls.
     */
    @Test
    void testCreation() {
        assertThrows(NullPointerException.class, () -> new GravatarProfileUrl(null, null));
        assertThrows(NullPointerException.class, () -> new GravatarProfileUrl("", null));
        assertThrows(IllegalArgumentException.class, () -> new GravatarProfileUrl("", ""));
        assertThrows(IllegalArgumentException.class, () -> new GravatarProfileUrl("plank", ""));
        assertDoesNotThrow(() -> new GravatarProfileUrl("plank", "plank"));
    }

    /**
     * Tests for the accessor and mutator methods.
     */
    @Test
    void testAccessorsMutators() {
        GravatarProfileUrl url = new GravatarProfileUrl("name", "link");
        assertEquals("name", url.getName());
        assertEquals("link", url.getLink());
    }

    /**
     * Tests for the to string method.
     */
    @Test
    void testToString() {
        GravatarProfileUrl url = new GravatarProfileUrl("name", "link");
        assertEquals("GravatarProfileUrl{name=\"name\", link=\"link\"}", url.toString());

        GravatarProfileUrl myGitHub =
                new GravatarProfileUrl("GitHub", "https://www.github.com/nathancheshire");
        assertEquals("GravatarProfileUrl{name=\"GitHub\", link=\"https://www.github.com/nathancheshire\"}",
                myGitHub.toString());
    }

    /**
     * Tests for the equals method.
     */
    @Test
    void testEquals() {
        GravatarProfileUrl url = new GravatarProfileUrl("name", "link");
        GravatarProfileUrl equalToUrl = new GravatarProfileUrl("name", "link");
        GravatarProfileUrl notEqualToUrl = new GravatarProfileUrl("other name", "link");
        GravatarProfileUrl notEqualToUrlEither = new GravatarProfileUrl("name", "other link");

        assertEquals(url, url);
        assertEquals(url, equalToUrl);
        assertNotEquals(equalToUrl, notEqualToUrl);
        assertNotEquals(equalToUrl, notEqualToUrlEither);
        assertNotEquals(url, new Object());
    }

    /**
     * Tests for the hash code method.
     */
    @Test
    void testHashCode() {
        GravatarProfileUrl url = new GravatarProfileUrl("name", "link");
        GravatarProfileUrl equalToUrl = new GravatarProfileUrl("name", "link");
        GravatarProfileUrl notEqualToUrl = new GravatarProfileUrl("other name", "other link");

        assertEquals(107906767, url.hashCode());
        assertEquals(107906767, equalToUrl.hashCode());
        assertEquals(890890447, notEqualToUrl.hashCode());
        assertEquals(url.hashCode(), equalToUrl.hashCode());
        assertNotEquals(equalToUrl.hashCode(), notEqualToUrl.hashCode());
    }
}

File: src/test/kotlin/com/github/natche/gravatarjavaclient/profile/serialization/GravatarProfileUrlTest.kt
package com.github.natche.gravatarjavaclient.profile.serialization

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

/**
 * Tests for [GravatarProfileUrl]s.
 */
class GravatarProfileUrlTest {
    /**
     * Tests for creation.
     */
    @Test
    fun testCreation() {
        assertThrows(NullPointerException::class.java) { GravatarProfileUrl(null, null) }
        assertThrows(NullPointerException::class.java) { GravatarProfileUrl("", null) }
        assertThrows(IllegalArgumentException::class.java) { GravatarProfileUrl("", "") }
        assertThrows(IllegalArgumentException::class.java) { GravatarProfileUrl("   ", "") }
        assertThrows(IllegalArgumentException::class.java) { GravatarProfileUrl("label", "") }
        assertThrows(IllegalArgumentException::class.java) { GravatarProfileUrl("label", "   ") }
        assertDoesNotThrow { GravatarProfileUrl("label", "url") }
    }

    /**
     * Tests for the accessors.
     */
    @Test
    fun testAccessors() {
        val url = GravatarProfileUrl("label", "url")
        assertEquals("label", url.label)
        assertEquals("url", url.url)
    }

    /**
     * Tests for the to string method.
     */
    @Test
    fun testToString() {
        val url = GravatarProfileUrl("label", "url")
        assertEquals("GravatarProfileUrl{label=\"label\", url=\"url\"}", url.toString())
    }

    /**
     * Tests for the hash code method.
     */
    @Test
    fun testHashCode() {
        val url = GravatarProfileUrl("label", "url")
        val equal = GravatarProfileUrl("label", "url")
        val notEqual = GravatarProfileUrl("other label", "url")

        assertEquals(-1110301445, url.hashCode())
        assertEquals(-1110301445, equal.hashCode())
        assertEquals(660405259, notEqual.hashCode())
        assertEquals(url.hashCode(), equal.hashCode())
        assertNotEquals(url.hashCode(), notEqual.hashCode())
    }

    /**
     * Tests for the equals method.
     */
    @Test
    fun testEquals() {
        val url = GravatarProfileUrl("label", "url")
        val equal = GravatarProfileUrl("label", "url")
        val differentUrl = GravatarProfileUrl("label", "other url")
        val differentLabel = GravatarProfileUrl("other label", "url")

        assertEquals(url, url)
        assertEquals(url, equal)
        assertNotEquals(url, differentUrl)
        assertNotEquals(url, differentLabel)
        assertNotEquals(url, Object())
    }
}